sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetReadSupport.scala

```diff
@@ -63,9 +74,22 @@ private[parquet] class ParquetReadSupport extends ReadSupport[UnsafeRow] with Lo
       StructType.fromString(schemaString)
     }
 
-    val parquetRequestedSchema =
+    val clippedParquetSchema =
       ParquetReadSupport.clipParquetSchema(context.getFileSchema, catalystRequestedSchema)
 
+    val parquetRequestedSchema = if (parquetMrCompatibility) {
+      // Parquet-mr will throw an exception if we try to read a superset of the file's schema.
+      // Therefore, we intersect our clipped schema with the underlying file's schema
+      ParquetReadSupport.intersectParquetGroups(clippedParquetSchema, context.getFileSchema)
+        .map(intersectionGroup =>
+          new MessageType(intersectionGroup.getName, intersectionGroup.getFields))
+        .getOrElse(ParquetSchemaConverter.EMPTY_MESSAGE)
+    } else {
+      // Spark's built-in Parquet reader will throw an exception in some cases if the requested
+      // schema is not the same as the clipped schema
```

Comment:
I think the built-in Parquet reader means vectorized reader. But I think we don't use `ParquetReadSupport` in vectorized reader?

Reply:
We do use it in the vectorized reader, but it's not easy to see. It's instantiated here: https://github.com/apache/spark/blob/38cec5f05066f75d868b64d360b04edb18dcebeb/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java#L141 via reflection.